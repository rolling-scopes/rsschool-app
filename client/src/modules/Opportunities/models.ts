import { ResumeDtoEnglishLevelEnum, ResumeDtoMilitaryServiceEnum } from 'api';
import { Dayjs } from 'dayjs';

export interface CourseData {
  locationName: string;
  courseFullName: string;
  isExpelled: boolean;
  certificateId: string | null;
  isCourseCompleted: boolean;
  totalScore: number;
  rank: number | null;
  mentor: {
    githubId: string;
    name: string;
  };
}

export type VisibleCourses = number[];

export interface CourseDataShortened {
  courseId: number;
  courseFullName: string;
}

export interface VisibleCoursesFormData {
  [id: string]: boolean;
}

export type ContactType = 'phone' | 'email' | 'skype' | 'telegram' | 'linkedin' | 'githubUsername' | 'website';

export interface UserData {
  avatarLink: string | null;
  name: string | null;
  desiredPosition: string | null;
  selfIntroLink: string | null;
  englishLevel: ResumeDtoEnglishLevelEnum | null;
  militaryService: ResumeDtoMilitaryServiceEnum | null;
  notes: string | null;
  startFrom: string | null;
  fullTime: boolean;
  locations: string | null;
}

export type Contacts = {
  [key in ContactType]: string | null;
};

export interface FieldData {
  name: string[];
  value: any;
  touched: boolean;
  validating: boolean;
  errors: string[];
}

export interface AllUserCVData extends Omit<UserData, 'uuid'>, Contacts {
  visibleCourses: VisibleCourses;
}

export interface UserDataToSubmit extends Omit<UserData, 'startFrom'> {
  startFrom: Dayjs;
  visibleCourses: VisibleCourses;
}

export interface AllDataToSubmit extends UserDataToSubmit, Contacts {}

export interface EditCVData extends AllUserCVData {
  expires: number | null;
  courses: CourseDataShortened[];
  visibleCourses: VisibleCourses;
}

export interface GetFullCVData extends AllUserCVData {
  expires: number | null;
  courses: CVStudentStats[];
  feedback: CVFeedback[];
}

export interface CVStudentStats {
  locationName: string;
  courseFullName: string;
  certificateId: string | null;
  isCourseCompleted: boolean;
  totalScore: number;
  rank: number | null;
  mentor: {
    githubId: string;
    name: string;
  };
}

export interface JobSeekerStudentStats extends CVStudentStats {
  courseName: string;
}

export interface JobSeekerData {
  name: string | null;
  desiredPosition: string | null;
  githubId: string;
  englishLevel: ResumeDtoEnglishLevelEnum | null;
  fullTime: boolean;
  locations: string | null;
  startFrom: string | null;
  courses: JobSeekerStudentStats[];
  feedback: JobSeekerFeedback[];
  expires: number;
  isHidden: boolean;
}

export interface CVFeedback {
  comment: string;
  feedbackDate: string;
}

export interface JobSeekerFeedback {
  badgeId: string;
}
