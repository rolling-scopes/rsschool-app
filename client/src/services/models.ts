import { Session } from 'components/withSession';
import {
  StudentBasic as CommonStudentBasic,
  DiscordServer,
  UserGroup,
  InterviewQuestion,
  InterviewQuestionCategory,
} from 'common/models';
import type { ProfileCourseDto } from 'api';

export interface Course extends ProfileCourseDto {}
export interface UpdateCourse extends ProfileCourseDto {
  disciplineId: number | null;
}

export interface StudentBasic extends CommonStudentBasic {}
export type { DiscordServer };
export type { UserGroup };
export type { InterviewQuestion };
export type { InterviewQuestionCategory };

export interface Mentor {
  lastName: string;
  firstName: string;
  githubId: string;

  studentId: number;
  userId: number;
  courseId: number;
}

export interface PageWithModalState<T> {
  data: T[];
  modalData: Partial<T> | null;
  modalAction: 'update' | 'create';
}

export interface CoursePageProps {
  session: Session;
  course: Course;
  params: Record<string, string>;
}

export type CourseOnlyPageProps = {
  course: Course;
  params?: Record<string, string>;
};
