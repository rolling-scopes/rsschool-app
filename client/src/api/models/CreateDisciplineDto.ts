/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */

export type CreateDisciplineDto = {
    name: string;
}
