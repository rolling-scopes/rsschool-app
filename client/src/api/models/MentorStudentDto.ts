/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */

export type MentorStudentDto = {
    name: string;
    id: number;
    cityName: string;
    countryName: string;
    githubId: string;
    totalScore: number;
    rank: number;
    hasFeedback: boolean;
}
