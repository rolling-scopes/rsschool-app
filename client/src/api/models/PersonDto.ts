/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */

export type PersonDto = {
    name: string;
    id: number;
}
