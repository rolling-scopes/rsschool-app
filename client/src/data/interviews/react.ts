import { InterviewTemplate } from './types';

export const reactTemplate: InterviewTemplate = {
  name: 'React interview',
  examplesUrl: 'https://github.com/rolling-scopes-school/tasks/blob/master/react/questions.md',
  categories: [
    {
      id: 1000,
      name: 'JSX',
      questions: [
        { id: 1001, name: 'What is JSX?' },
        { id: 1002, name: 'Is it possible to use React without JSX?' },
      ],
    },
    {
      id: 1010,
      name: 'Styled-components',
      questions: [{ id: 1011, name: 'Using styled-components' }],
    },
    {
      id: 1020,
      name: 'Components',
      questions: [
        { id: 1021, name: 'Component Definition (Class-based, Functional)' },
        { id: 1022, name: 'Components Composition' },
      ],
    },
    {
      id: 1030,
      name: 'React DOM',
      questions: [
        { id: 1031, name: 'What is the virtual DOM? How does react use the virtual DOM to render the UI?' },
        { id: 1032, name: 'Is the virtual DOM the same as the shadow DOM?' },
        { id: 1033, name: 'What is the difference between the virtual DOM and the real DOM?' },
      ],
    },
    {
      id: 1040,
      name: 'Render',
      questions: [
        { id: 1041, name: 'When is a component rendered?' },
        { id: 1042, name: 'How not to render on props change?' },
        { id: 1043, name: 'Is it OK to use arrow functions in render methods?' },
        { id: 1044, name: 'Components rendering system' },
      ],
    },
    {
      id: 1050,
      name: 'Interaction between components',
      questions: [
        { id: 1051, name: 'How do you pass a value from parent to child?' },
        { id: 1052, name: 'How do you pass a value from child to parent?' },
        { id: 1053, name: 'What is prop drilling?' },
        { id: 1054, name: 'Can a child component modify its own props?' },
        { id: 1055, name: 'How do you pass a value from sibling to sibling?' },
      ],
    },
    {
      id: 1060,
      name: 'Lifecycle and State',
      questions: [
        { id: 1061, name: 'What is the difference between props and state?' },
        { id: 1062, name: 'How does state in a class component differ from state in a functional component?' },
        { id: 1063, name: 'What is the component lifecycle?' },
        { id: 1064, name: 'How do you update lifecycle in function components?' },
        { id: 1065, name: 'Controlled/uncontrolled components' },
        { id: 1066, name: 'Stateful vs stateless components' },
      ],
    },
    {
      id: 1070,
      name: 'Ref',
      questions: [
        { id: 1071, name: 'What is the difference between refs and state variables?' },
        { id: 1072, name: 'When is the best time to use refs?' },
        { id: 1073, name: 'What is the proper way to update a ref in a function component?' },
      ],
    },
    {
      id: 1080,
      name: 'Context',
      questions: [
        { id: 1081, name: 'What is the difference between the context API and prop drilling?' },
        { id: 1082, name: "When shouldn't you use the context API?" },
      ],
    },
    {
      id: 1090,
      name: 'Redux',
      questions: [
        { id: 1091, name: 'Enumerate base principles' },
        { id: 1092, name: 'What is the typical flow of data in a React + Redux app?' },
        { id: 1093, name: 'Benefits of Redux?' },
        { id: 1094, name: 'Async Redux flow' },
      ],
    },
    {
      id: 1100,
      name: 'State Management by managers',
      questions: [
        { id: 1101, name: 'Mobx' },
        { id: 1102, name: 'Reflux' },
      ],
    },
    {
      id: 1110,
      name: 'Routing',
      questions: [
        { id: 1111, name: 'React Router' },
        { id: 1112, name: 'History' },
      ],
    },
    {
      id: 1120,
      name: 'Other',
      questions: [
        { id: 1121, name: 'Is it a good idea to use Math.random for keys?' },
        { id: 1122, name: 'What are the limitations of React?' },
        { id: 1123, name: 'What is a higher order component?' },
        { id: 1124, name: 'What are uncontrolled and controlled components?' },
        { id: 1125, name: 'React optimizations' },
      ],
    },
    {
      id: 1130,
      name: 'Coding task',
      questions: [{ id: 1131, name: 'Small react app: form, button, results list' }],
    },
    {
      id: 1140,
      name: 'Next.js',
      questions: [
        { id: 1141, name: 'What is Next.js, and how does it differ from traditional React applications?' },
        { id: 1142, name: 'Explain the concept of server-side rendering (SSR) in Next.js' },
        { id: 1143, name: 'How does Next.js handle client-side routing?' },
        { id: 1144, name: 'What are the benefits of using Next.js for building React applications?' },
        {
          id: 1145,
          name: 'What is the purpose of the getInitialProps function in Next.js? How is it different from using getStaticProps or getServerSideProps?',
        },
        { id: 1146, name: 'How does Next.js handle automatic code splitting, and why is it important?' },
        {
          id: 1147,
          name: 'What is the purpose of the Link component in Next.js, and how does it differ from traditional anchor (<a>) tags?',
        },
        { id: 1148, name: 'How can you configure routing in a Next.js application?' },
      ],
    },
    {
      id: 1150,
      name: 'Data Fetching',
      questions: [
        {
          id: 1151,
          name: 'Compare and contrast getStaticProps and getServerSideProps in terms of use cases and performance considerations',
        },
        {
          id: 1152,
          name: 'When would you choose to use getStaticPaths in Next.js, and how does it relate to dynamic routes?',
        },
      ],
    },
    {
      id: 1160,
      name: 'Advanced Topics',
      questions: [
        { id: 1161, name: 'Describe the purpose and use cases of API routes in Next.js' },
        { id: 1162, name: 'Explain the concept of middleware in Next.js and its role in the application lifecycle' },
        { id: 1163, name: 'How does Next.js handle authentication in applications?' },
      ],
    },
  ],
};
