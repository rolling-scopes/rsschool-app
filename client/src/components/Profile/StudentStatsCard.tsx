import * as React from 'react';
import isEqual from 'lodash/isEqual';
import random from 'lodash/random';
import { Typography, List, Button, Progress, Modal, Input, Divider } from 'antd';
import CommonCard from './CommonCard';
import StudentStatsModal from './StudentStatsModal';
import { StudentStats } from '../../../../common/models/profile';
import { ConfigurableProfilePermissions } from '../../../../common/models/profile';
import { ChangedPermissionsSettings } from 'pages/profile';
import { CheckboxChangeEvent } from 'antd/lib/checkbox';
import { BookOutlined, FullscreenOutlined, SafetyCertificateTwoTone } from '@ant-design/icons';
import { CourseService } from '../../services/course';
import { WarningTwoTone } from '@ant-design/icons';
import { CSSProperties } from 'react';

const { Text, Paragraph } = Typography;

type Props = {
  data: StudentStats[];
  isEditingModeEnabled: boolean;
  isProfileOwner: boolean;
  permissionsSettings?: ConfigurableProfilePermissions;
  onPermissionsSettingsChange: (event: CheckboxChangeEvent, settings: ChangedPermissionsSettings) => void;
  username: string;
};

type State = {
  courseIndex: number;
  coursesProgress: number[];
  scoredTasks: number[];
  isStudentStatsModalVisible: boolean;
  isExpelConfirmationModalVisible: boolean;
};

class StudentStatsCard extends React.Component<Props, State> {
  state = {
    courseIndex: 0,
    coursesProgress: [],
    scoredTasks: [],
    isStudentStatsModalVisible: false,
    isExpelConfirmationModalVisible: false,
  };

  shouldComponentUpdate = (nextProps: Props, nextState: State) =>
    !isEqual(
      nextProps.permissionsSettings?.isStudentStatsVisible,
      this.props.permissionsSettings?.isStudentStatsVisible,
    ) ||
    !isEqual(nextProps.isEditingModeEnabled, this.props.isEditingModeEnabled) ||
    !isEqual(nextState.isStudentStatsModalVisible, this.state.isStudentStatsModalVisible) ||
    !isEqual(nextState.coursesProgress, this.state.coursesProgress);

  private filterPermissions = ({ isStudentStatsVisible }: Partial<ConfigurableProfilePermissions>) => ({
    isStudentStatsVisible,
  });

  private showStudentStatsModal = (courseIndex: number) => {
    this.setState({ courseIndex, isStudentStatsModalVisible: true });
  };

  private showExpelConfirmationModal = (gitHubId: string, courseId: number) => {
    const { isExpelConfirmationModalVisible } = this.state;

    const keyLength = 8;

    let key = '';

    for (let i = 0; i < keyLength; i++) key += random(0, 9);

    const title = (
      <Typography.Title level={3} style={{ textAlign: 'center' }}>
        <WarningTwoTone twoToneColor="#fcbe03" /> <Text strong>Are you sure?</Text>
        <WarningTwoTone twoToneColor="#fcbe03" />
      </Typography.Title>
    );

    const checkKeyMatch = (e: any) => {
      if (e.target.value === key) {
        modal.update({
          okButtonProps: { disabled: false },
        });
      } else {
        modal.update({
          okButtonProps: { disabled: true },
        });
      }
    };

    const message =
      "Are you sure you want to leave the course? Your learning will be finished. You won't be able to return to the course on your own.";
    const messageRu =
      'Вы уверены, что хотите покинуть курс? Ваше обучение будет окончено. Вы не сможете вернуться на курс самостоятельно.';

    const textStyle: CSSProperties = { textAlign: 'center' };

    const content = (
      <>
        <Paragraph style={textStyle} underline strong>
          {message}
        </Paragraph>
        <Paragraph style={textStyle} underline strong>
          {messageRu}
        </Paragraph>
        <Divider plain style={{ whiteSpace: 'normal' }}>
          Enter following number to confirm action: <Text strong>{key}</Text>
        </Divider>
        <Input placeholder="Enter the number" type="text" onChange={checkKeyMatch} />
      </>
    );

    const modal = Modal.confirm({
      maskStyle: { backgroundColor: 'red' },
      icon: null,
      title: title,
      content: content,
      centered: true,
      onOk: () => this.selfExpelStudent(gitHubId, courseId),
      visible: isExpelConfirmationModalVisible,
      onCancel: () => this.hideExpelConfirmationModal(),
      okButtonProps: { disabled: true },
      maskClosable: true,
    });
  };

  private hideStudentStatsModal = () => {
    this.setState({ isStudentStatsModalVisible: false });
  };

  private hideExpelConfirmationModal = () => {
    this.setState({ isExpelConfirmationModalVisible: false });
  };

  private selfExpelStudent = async (gitHubId: string, courseId: number) => {
    const courseService = new CourseService(courseId);
    const result = await courseService.selfExpel(gitHubId, courseId.toString());
    if (result.status === 200) {
      window.location.reload();
    }
  };

  private countScoredTasks = (tasks: { score: number }[]) => tasks.filter(({ score }) => score !== null).length;
  private countCourseCompletionPercentage = (tasks: { score: number }[]) =>
    Number(((tasks.filter(({ score }) => score !== null).length / tasks.length) * 100).toFixed(1));

  componentDidMount() {
    const stats = this.props.data;
    const scoredTasks = stats.map(({ tasks }) => this.countScoredTasks(tasks));
    const coursesProgress = stats.map(({ tasks }) => this.countCourseCompletionPercentage(tasks));
    this.setState({ coursesProgress, scoredTasks });
  }

  render() {
    const { isEditingModeEnabled, permissionsSettings, onPermissionsSettingsChange, isProfileOwner } = this.props;
    const stats = this.props.data;
    const gitHubId: string = this.props.username;
    const { isStudentStatsModalVisible, courseIndex, coursesProgress, scoredTasks } = this.state;
    return (
      <>
        <StudentStatsModal
          stats={stats[courseIndex]}
          courseProgress={coursesProgress[courseIndex]}
          scoredTasks={scoredTasks[courseIndex]}
          isVisible={isStudentStatsModalVisible}
          onHide={this.hideStudentStatsModal}
        />
        <CommonCard
          title="Student Statistics"
          icon={<BookOutlined />}
          content={
            <List
              itemLayout="horizontal"
              dataSource={stats}
              renderItem={(
                {
                  courseName,
                  locationName,
                  mentor,
                  totalScore,
                  isExpelled,
                  position,
                  isCourseCompleted,
                  certificateId,
                  courseId,
                },
                idx,
              ) => {
                const isActive = !(isExpelled || isCourseCompleted);
                return (
                  <List.Item style={{ display: 'flex', justifyContent: 'space-between' }}>
                    <div style={{ flexGrow: 2 }}>
                      <p style={{ marginBottom: 0 }}>
                        <Text strong>
                          {courseName}
                          {locationName && ` / ${locationName}`}
                        </Text>
                      </p>
                      <div style={{ width: '80%', marginBottom: 5 }}>
                        <Progress
                          percent={coursesProgress.length ? coursesProgress[idx] : 0}
                          status={isExpelled ? 'exception' : isCourseCompleted ? 'success' : undefined}
                          size="small"
                        />
                      </div>
                      {certificateId && (
                        <p style={{ fontSize: 16, marginBottom: 5 }}>
                          <SafetyCertificateTwoTone twoToneColor="#52c41a" />{' '}
                          <a target="__blank" href={`/certificate/${certificateId}`}>
                            Certificate
                          </a>
                        </p>
                      )}
                      {mentor.githubId && (
                        <p style={{ fontSize: 12, marginBottom: 5 }}>
                          Mentor: <a href={`/profile?githubId=${mentor.githubId}`}>{mentor.name}</a>
                        </p>
                      )}
                      {position && (
                        <p style={{ fontSize: 12, marginBottom: 5 }}>
                          Position: <Text strong>{position}</Text>
                        </p>
                      )}
                      <p style={{ fontSize: 12, marginBottom: 5 }}>
                        Score: <Text mark>{totalScore}</Text>
                      </p>

                      {isActive && isProfileOwner ? (
                        <Button danger size="small" onClick={() => this.showExpelConfirmationModal(gitHubId, courseId)}>
                          Leave course
                        </Button>
                      ) : (
                        ''
                      )}
                    </div>
                    <Button type="dashed" onClick={this.showStudentStatsModal.bind(null, idx)}>
                      <FullscreenOutlined />
                    </Button>
                  </List.Item>
                );
              }}
            />
          }
          permissionsSettings={permissionsSettings ? this.filterPermissions(permissionsSettings) : undefined}
          isEditingModeEnabled={isEditingModeEnabled}
          onPermissionsSettingsChange={onPermissionsSettingsChange}
        />
      </>
    );
  }
}

export default StudentStatsCard;
