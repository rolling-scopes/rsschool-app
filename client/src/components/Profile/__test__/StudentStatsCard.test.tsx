import React from 'react';
import { render } from '@testing-library/react';
import StudentStatsCard from '../StudentStatsCard';
import { StudentStats } from 'common/models/profile';

describe('', () => {
  const githubId = 'test';
  const data = [
    {
      courseId: 1,
      courseName: 'rs-2018-q1',
      locationName: 'Minsk',
      courseFullName: 'Rolling Scopes School 2018 Q1',
      isExpelled: false,
      expellingReason: '',
      isCourseCompleted: true,
      totalScore: 1201,
      rank: 32,
      mentor: {
        githubId: 'andrew123',
        name: 'Andrey Andreev',
      },
      tasks: [
        {
          maxScore: 130,
          scoreWeight: 1,
          name: 'Task 1',
          descriptionUri: 'https://description.com',
          taskGithubPrUris: 'https://description.com',
          score: 120,
          comment: 'test',
        },
        {
          maxScore: 100,
          scoreWeight: 1,
          name: 'Task 2',
          descriptionUri: 'https://description.com',
          taskGithubPrUris: 'https://description.com',
          score: 20,
          comment: 'test',
        },
        {
          maxScore: 110,
          scoreWeight: 1,
          name: 'Task 3',
          descriptionUri: 'https://description.com',
          taskGithubPrUris: 'https://description.com',
          score: 90,
          comment: 'test',
        },
      ],
    },
    {
      courseId: 1,
      courseName: 'rs-2019-q1',
      locationName: 'Minsk',
      courseFullName: 'Rolling Scopes School 2019 Q1',
      isExpelled: true,
      expellingReason: 'test',
      isCourseCompleted: false,
      totalScore: 101,
      rank: 32,
      mentor: {
        githubId: 'dimon12',
        name: 'Dima Testovich',
      },
      tasks: [
        {
          maxScore: 100,
          scoreWeight: 1,
          name: 'Task 1',
          descriptionUri: 'https://description.com',
          taskGithubPrUris: 'https://description.com',
          score: 20,
          comment: 'test',
        },
        {
          maxScore: 100,
          scoreWeight: 1,
          name: 'Task 2',
          descriptionUri: 'https://description.com',
          taskGithubPrUris: 'https://description.com',
          score: null,
          comment: null,
        },
        {
          maxScore: 100,
          scoreWeight: 1,
          name: 'Task 3',
          descriptionUri: 'https://description.com',
          taskGithubPrUris: 'https://description.com',
          score: 10,
          comment: 'test',
        },
        {
          maxScore: 100,
          scoreWeight: 1,
          name: 'Task 4',
          descriptionUri: 'https://description.com',
          taskGithubPrUris: 'https://description.com',
          score: null,
          comment: null,
        },
      ],
    },
  ] as StudentStats[];

  it('should render correctly', () => {
    const output = render(<StudentStatsCard isProfileOwner={false} data={data} username={githubId} />);
    expect(output.container).toMatchSnapshot();
  });
});
