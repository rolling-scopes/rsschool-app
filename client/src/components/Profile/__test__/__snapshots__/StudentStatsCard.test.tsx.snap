// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentStatsCard should render correctly 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered css-dev-only-do-not-override-w8mnev"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          <h2
            class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-w8mnev"
            style="font-size: 16px; margin-bottom: 0px; display: flex; align-items: center; justify-content: space-between;"
          >
            <span>
              <span
                aria-label="book"
                class="anticon anticon-book"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="book"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M832 64H192c-17.7 0-32 14.3-32 32v832c0 17.7 14.3 32 32 32h640c17.7 0 32-14.3 32-32V96c0-17.7-14.3-32-32-32zm-260 72h96v209.9L621.5 312 572 347.4V136zm220 752H232V136h280v296.9c0 3.3 1 6.6 3 9.3a15.9 15.9 0 0022.3 3.7l83.8-59.9 81.4 59.4c2.7 2 6 3.1 9.4 3.1 8.8 0 16-7.2 16-16V136h64v752z"
                  />
                </svg>
              </span>
               
              Student Statistics
            </span>
          </h2>
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <div
        class="ant-list ant-list-split css-dev-only-do-not-override-w8mnev"
      >
        <div
          class="ant-spin-nested-loading css-dev-only-do-not-override-w8mnev"
        >
          <div
            class="ant-spin-container"
          >
            <ul
              class="ant-list-items"
            >
              <li
                class="ant-list-item"
                style="display: flex; justify-content: space-between;"
              >
                <div
                  style="flex-grow: 2;"
                >
                  <p
                    style="margin-bottom: 0px;"
                  >
                    <span
                      class="ant-typography css-dev-only-do-not-override-w8mnev"
                    >
                      <strong>
                        rs-2018-q1
                         / Minsk
                      </strong>
                    </span>
                  </p>
                  <div
                    style="width: 80%; margin-bottom: 5px;"
                  >
                    <div
                      class="ant-progress ant-progress-line ant-progress-status-success ant-progress-show-info ant-progress-small css-dev-only-do-not-override-w8mnev"
                      role="progressbar"
                    >
                      <div
                        class="ant-progress-outer"
                        style="width: 100%; height: 6px;"
                      >
                        <div
                          class="ant-progress-inner"
                        >
                          <div
                            class="ant-progress-bg"
                            style="width: 100%; height: 6px;"
                          />
                        </div>
                      </div>
                      <span
                        class="ant-progress-text"
                      >
                        <span
                          aria-label="check-circle"
                          class="anticon anticon-check-circle"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="check-circle"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm193.5 301.7l-210.6 292a31.8 31.8 0 01-51.7 0L318.5 484.9c-3.8-5.3 0-12.7 6.5-12.7h46.9c10.2 0 19.9 4.9 25.9 13.3l71.2 98.8 157.2-218c6-8.3 15.6-13.3 25.9-13.3H699c6.5 0 10.3 7.4 6.5 12.7z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                  </div>
                  <p
                    style="font-size: 12px; margin-bottom: 5px;"
                  >
                    Mentor: 
                    <a
                      href="/profile?githubId=andrew123"
                    >
                      Andrey Andreev
                    </a>
                  </p>
                  <p
                    style="font-size: 12px; margin-bottom: 5px;"
                  >
                    Position: 
                    32
                  </p>
                  <p
                    style="font-size: 12px; margin-bottom: 5px;"
                  >
                    Score: 
                    1201
                  </p>
                </div>
                <button
                  class="ant-btn css-dev-only-do-not-override-w8mnev ant-btn-dashed"
                  type="button"
                >
                  <span
                    aria-label="fullscreen"
                    class="anticon anticon-fullscreen"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="fullscreen"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M290 236.4l43.9-43.9a8.01 8.01 0 00-4.7-13.6L169 160c-5.1-.6-9.5 3.7-8.9 8.9L179 329.1c.8 6.6 8.9 9.4 13.6 4.7l43.7-43.7L370 423.7c3.1 3.1 8.2 3.1 11.3 0l42.4-42.3c3.1-3.1 3.1-8.2 0-11.3L290 236.4zm352.7 187.3c3.1 3.1 8.2 3.1 11.3 0l133.7-133.6 43.7 43.7a8.01 8.01 0 0013.6-4.7L863.9 169c.6-5.1-3.7-9.5-8.9-8.9L694.8 179c-6.6.8-9.4 8.9-4.7 13.6l43.9 43.9L600.3 370a8.03 8.03 0 000 11.3l42.4 42.4zM845 694.9c-.8-6.6-8.9-9.4-13.6-4.7l-43.7 43.7L654 600.3a8.03 8.03 0 00-11.3 0l-42.4 42.3a8.03 8.03 0 000 11.3L734 787.6l-43.9 43.9a8.01 8.01 0 004.7 13.6L855 864c5.1.6 9.5-3.7 8.9-8.9L845 694.9zm-463.7-94.6a8.03 8.03 0 00-11.3 0L236.3 733.9l-43.7-43.7a8.01 8.01 0 00-13.6 4.7L160.1 855c-.6 5.1 3.7 9.5 8.9 8.9L329.2 845c6.6-.8 9.4-8.9 4.7-13.6L290 787.6 423.7 654c3.1-3.1 3.1-8.2 0-11.3l-42.4-42.4z"
                      />
                    </svg>
                  </span>
                </button>
              </li>
              <li
                class="ant-list-item"
                style="display: flex; justify-content: space-between;"
              >
                <div
                  style="flex-grow: 2;"
                >
                  <p
                    style="margin-bottom: 0px;"
                  >
                    <span
                      class="ant-typography css-dev-only-do-not-override-w8mnev"
                    >
                      <strong>
                        rs-2019-q1
                         / Minsk
                      </strong>
                    </span>
                  </p>
                  <div
                    style="width: 80%; margin-bottom: 5px;"
                  >
                    <div
                      class="ant-progress ant-progress-line ant-progress-status-exception ant-progress-show-info ant-progress-small css-dev-only-do-not-override-w8mnev"
                      role="progressbar"
                    >
                      <div
                        class="ant-progress-outer"
                        style="width: 100%; height: 6px;"
                      >
                        <div
                          class="ant-progress-inner"
                        >
                          <div
                            class="ant-progress-bg"
                            style="width: 50%; height: 6px;"
                          />
                        </div>
                      </div>
                      <span
                        class="ant-progress-text"
                      >
                        <span
                          aria-label="close-circle"
                          class="anticon anticon-close-circle"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="close-circle"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                  </div>
                  <p
                    style="font-size: 12px; margin-bottom: 5px;"
                  >
                    Mentor: 
                    <a
                      href="/profile?githubId=dimon12"
                    >
                      Dima Testovich
                    </a>
                  </p>
                  <p
                    style="font-size: 12px; margin-bottom: 5px;"
                  >
                    Position: 
                    32
                  </p>
                  <p
                    style="font-size: 12px; margin-bottom: 5px;"
                  >
                    Score: 
                    101
                  </p>
                </div>
                <button
                  class="ant-btn css-dev-only-do-not-override-w8mnev ant-btn-dashed"
                  type="button"
                >
                  <span
                    aria-label="fullscreen"
                    class="anticon anticon-fullscreen"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="fullscreen"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M290 236.4l43.9-43.9a8.01 8.01 0 00-4.7-13.6L169 160c-5.1-.6-9.5 3.7-8.9 8.9L179 329.1c.8 6.6 8.9 9.4 13.6 4.7l43.7-43.7L370 423.7c3.1 3.1 8.2 3.1 11.3 0l42.4-42.3c3.1-3.1 3.1-8.2 0-11.3L290 236.4zm352.7 187.3c3.1 3.1 8.2 3.1 11.3 0l133.7-133.6 43.7 43.7a8.01 8.01 0 0013.6-4.7L863.9 169c.6-5.1-3.7-9.5-8.9-8.9L694.8 179c-6.6.8-9.4 8.9-4.7 13.6l43.9 43.9L600.3 370a8.03 8.03 0 000 11.3l42.4 42.4zM845 694.9c-.8-6.6-8.9-9.4-13.6-4.7l-43.7 43.7L654 600.3a8.03 8.03 0 00-11.3 0l-42.4 42.3a8.03 8.03 0 000 11.3L734 787.6l-43.9 43.9a8.01 8.01 0 004.7 13.6L855 864c5.1.6 9.5-3.7 8.9-8.9L845 694.9zm-463.7-94.6a8.03 8.03 0 00-11.3 0L236.3 733.9l-43.7-43.7a8.01 8.01 0 00-13.6 4.7L160.1 855c-.6 5.1 3.7 9.5 8.9 8.9L329.2 845c6.6-.8 9.4-8.9 4.7-13.6L290 787.6 423.7 654c3.1-3.1 3.1-8.2 0-11.3l-42.4-42.4z"
                      />
                    </svg>
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
